File: example/modules/frame-stats/android/src/main/java/expo/modules/framestats/FrameStatsModule.kt
package expo.modules.framestats

import android.os.Build
import android.os.Debug
import android.util.Log
import android.view.Display
import android.view.WindowManager
import expo.modules.kotlin.functions.Queues
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

/**
 * Exposes [FrameRecorder] to JS. Recording starts and stops on the main thread
 * because that is the thread whose frame intervals are measured.
 */
class FrameStatsModule : Module() {
  private var recorder: FrameRecorder? = null

  override fun definition() = ModuleDefinition {
    Name("FrameStats")

    AsyncFunction("setKeepAwake") { enabled: Boolean ->
      val window = appContext.currentActivity?.window
      if (enabled) window?.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
      else window?.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("start") { _: String ->
      recorder?.stop()
      recorder = FrameRecorder(::currentDisplay).also { it.start() }
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("stop") {
      val active = recorder
      recorder = null
      active?.stop() ?: FrameRecorder.emptyRecording(currentDisplay())
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("memoryFootprint") {
      // Proportional set size in bytes: the closest Android equivalent of
      // iOS `phys_footprint`.
      Debug.getPss().toDouble() * 1024.0
    }

    AsyncFunction("displayRefreshRate") {
      (currentDisplay()?.refreshRate ?: 60f).toDouble()
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("logLine") { line: String ->
      val context = requireNotNull(appContext.reactContext)
      File(context.filesDir, "marker-view-benchmark.jsonl").appendText(line + "\n")
      Log.i("NitroMapsBenchmark", line)
    }
  }

  private fun currentDisplay(): Display? {
    val activity = appContext.currentActivity ?: return null
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
      activity.display
    } else {
      @Suppress("DEPRECATION")
      activity.windowManager.defaultDisplay
    }
  }
}

File: example/modules/frame-stats/ios/FrameStatsModule.swift
import ExpoModulesCore
import UIKit

/// Exposes `FrameRecorder` to JS. Recording starts and stops on the main
/// thread because that is the thread whose frame intervals are measured.
public final class FrameStatsModule: Module {
  private var recorder: FrameRecorder?
  private var initialEnvironment: [String: Any] = [:]

  private func environment() -> [String: Any] {
    var mounted = 0
    var visible = 0
    func visit(_ view: UIView, hidden: Bool) {
      let hidden = hidden || view.isHidden || view.alpha == 0
      if NSStringFromClass(type(of: view)).hasSuffix(".NitroMarkerContentView") {
        mounted += 1
        if !hidden { visible += 1 }
      }
      for child in view.subviews { visit(child, hidden: hidden) }
    }
    for scene in UIApplication.shared.connectedScenes.compactMap({ $0 as? UIWindowScene }) {
      for window in scene.windows { visit(window, hidden: false) }
    }
    return [
      "os": UIDevice.current.systemVersion,
      "thermalState": ProcessInfo.processInfo.thermalState.rawValue,
      "lowPowerMode": ProcessInfo.processInfo.isLowPowerModeEnabled,
      "mountedMarkerViews": mounted,
      "visibleMarkerViews": visible,
      "applicationActive": UIApplication.shared.applicationState == .active,
    ]
  }

  public func definition() -> ModuleDefinition {
    Name("FrameStats")

    AsyncFunction("setKeepAwake") { (enabled: Bool) -> Void in
      UIApplication.shared.isIdleTimerDisabled = enabled
    }.runOnQueue(.main)

    AsyncFunction("start") { (label: String) -> Void in
      self.recorder?.stop()
      self.initialEnvironment = self.environment()
      let recorder = FrameRecorder()
      recorder.start(label: label)
      self.recorder = recorder
    }.runOnQueue(.main)

    AsyncFunction("stop") { () -> [String: Any] in
      guard let recorder = self.recorder else {
        return FrameRecorder.emptyRecording()
      }
      self.recorder = nil
      var result = recorder.stop()
      result["environmentBefore"] = self.initialEnvironment
      result["environmentAfter"] = self.environment()
      return result
    }.runOnQueue(.main)

    AsyncFunction("memoryFootprint") { () -> Double in
      Double(FrameRecorder.memoryFootprintBytes())
    }

    AsyncFunction("displayRefreshRate") { () -> Double in
      Double(UIScreen.main.maximumFramesPerSecond)
    }.runOnQueue(.main)

    AsyncFunction("logLine") { (line: String) throws -> Void in
      let documents = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
      let url = documents.appendingPathComponent("marker-view-benchmark.jsonl")
      if !FileManager.default.fileExists(atPath: url.path) {
        FileManager.default.createFile(atPath: url.path, contents: nil)
      }
      let file = try FileHandle(forWritingTo: url)
      defer { try? file.close() }
      try file.seekToEnd()
      try file.write(contentsOf: Data((line + "\n").utf8))
      NSLog("%@", line)
    }
  }
}
